feat(cli): allow running `projen` in a subdirectory

fynnfluegge opened the pull request:
This PR adds the feature to run `projen` commands from subdirectories in a projen project.

#### Closes #2418.
---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mrgrain reviewed (changes requested):
I have a feeling if you add a test with a `.projenrc.ts` or `.projenrc.py` file this will not work.

fynnfluegge commented:
> I have a feeling if you add a test with a `.projenrc.ts` or `.projenrc.py` file this will not work.

True, I think I should rather search for toplevel `.projen` directory as you mentioned before

fynnfluegge commented:
Hey @mrgrain I modified `findProjenRcJs` to search for `ts`, `py`, `java`, `json` files, too. The approach to search for `.projen` directory worked also but the problem was if you run `npx projen` from root in a subproject it fails, since you need to find the most upper `.projen`. Also possible to traverse through all directories and return the last found `.projen` directory but I don't like that 😄
With `yarn link` I was able to run `npx projen` from any directory in my projen project with this PR. What's eventually missing are a few more tests.

mrgrain commented:
> Hey @mrgrain I modified `findProjenRcJs` to search for `ts`, `py`, `java`, `json` files, too.
That is definitely and improvement! Although some projects allow you to choose any arbitrary filename. 🤔 

>The approach to search for `.projen` directory worked also but the problem was if you run `npx projen` from root in a subproject it fails, since you need to find the most upper `.projen`. Also possible to traverse through all directories and return the last found `.projen` directory but I don't like that, what's your opinion? 😄 With `yarn link` I was able to run `npx projen` from any directory in my projen project with this PR. What's eventually missing are a few more tests.

Okay, feels like we are slowly getting to the core of the problem. I suppose part of the question is: What should happen when I run projen commands in a subproject? There seems to be a difference in running projen from a subdirectory of a project and running projen from a subproject.

At the moment I can `cd subproject && npx projen build` and it would only build this subproject. I'm not sure we can just take this away. What doesn't seems to work in the subproject directory is the default task. So maybe we need to fix that?

Let me set up an example repo.

fynnfluegge commented:
> At the moment I can `cd subproject && npx projen build` and it would only build this subproject. I'm not sure we can just take this away. What doesn't seems to work in the subproject directory is the default task. So maybe we need to fix that?
> 
> Let me set up an example repo.
> 
> Edit: Maybe something like this needs to be the default task for subprojects [mrgrain/projen-subproject-test@6e32206](https://github.com/mrgrain/projen-subproject-test/commit/6e322062079e56a726a645122953cf972fddb4bf)

Alright, guess I got it. So what we want is, when running `npx projen` in a subproject that this command should only run in the subproject. Hence, searching for `.projen` directory is the way as you initially mentioned. 

I was just confused about the failure of the `default` task in the root of a subproject and thought this should be handled by executing it from the parent project root and I switched back to search for the `.projenrc` to fix this. But if we say this is a bug this can be fixed separately and I change my PR back to searching for `.projen` which is really straight forward. 👍

mrgrain commented:
> Alright, guess I got it. So what we want is, when running `npx projen` in a subproject that this command should only run in the subproject. Hence, searching for `.projen` directory is the way as you initially mentioned.
> 
> I was just confused about the failure of the `default` task in the root of a subproject and thought this should be handled by executing it from the parent project root and I switched back to search for the `.projenrc` to fix this.

I mean I got confused too. Glad you started thinking about subprojects 😅 


> But if we say this is a bug this can be fixed separately and I change my PR back to searching for `.projen` which is really straight forward. 👍 What do you think?

Yes, that sounds like a good plan to me. We might have to sequence the changes correctly, i.e. first merging the fix for the default task.

fynnfluegge commented:
> Yes, that sounds like a good plan to me. We might have to sequence the changes correctly, i.e. first merging the fix for the default task.

Alright, created an issue for this, let's first fix #2497 ✅
I pushed a new commit that passes the directory of `.projen` the `TaskRuntime` so that projen commands can be executed an any subdirectory. But let's see how this will work out together with the fix for the default task in subprojects.

mrgrain commented on src/cli/index.ts:
```diff
@@ -35,7 +37,7 @@ async function main() {
   ya.options("rc", {
     desc: "path to .projenrc.js file",
-    default: DEFAULT_RC,
+    default: DEFAULT_RC ?? PROJEN_RC,
```
```suggestion
    default: DEFAULT_RC,
```

fynnfluegge replied:
oops, forgot to undo this

mrgrain commented on src/cli/index.ts:
```diff
@@ -81,3 +83,20 @@ main().catch((e) => {
   process.exit(1);
 });
+
+/** Run up project tree to find .projen directory
+ *
```
```suggestion
/**
 * Run up project tree to find .projen directory
```

mrgrain commented on src/cli/index.ts:
```diff
@@ -81,3 +83,20 @@ main().catch((e) => {
+ * @returns path to .projen directory or undefined if not found
+ */
+function findProjenDir(cwd: string): string | undefined {
+  if (cwd === "/") {
+    return undefined;
+  }
+
+  if (fs.existsSync(resolve(cwd, PROJEN_DIR))) {
+    return cwd;
+  }
```
I think technically these need to be switched around if you want to allow `/.projen`
```suggestion
  if (fs.existsSync(resolve(cwd, PROJEN_DIR))) {
    return cwd;
  }

  if (cwd === "/") {
    return undefined;
  }
```

mrgrain commented on src/task-runtime.ts:
```diff
@@ -35,7 +35,7 @@ export class TaskRuntime {
 
   constructor(workdir: string) {
-    this.workdir = resolve(workdir);
+    this.workdir = workdir;
```
Let's keep other changes to a minimum. Who knows where else we use `TaskRuntime` with a relative path.
```suggestion
    this.workdir = resolve(workdir);
```

fynnfluegge replied:
I agree 👍

mrgrain reviewed (changes requested):
Also needs tests 🚀

fynnfluegge commented:
> Also needs tests 🚀

Will add some tests asap ✅

fynnfluegge commented:
Added a test that runs task in subdirectory ✅🚀

mrgrain reviewed (changes requested):
I reckon we need to more test cases here:

- In a nested subdirectory, projen finds the closest `.projen/task.json` and not the root one
- If there is no `.projen/task.json` file

Also - any chance these could use real projects and subprojects instead of mocks?

fynnfluegge commented:
> Also - any chance these could use real projects and subprojects instead of mocks?

Sure ✅

> In a nested subdirectory, projen finds the closest .projen/task.json and not the root one

Added ✅

> If there is no .projen/task.json file

Added ✅ But not sure if I got what you mean exactly 😄 That it terminates with an error and don't run in an infinite loop for example?

mrgrain commented:
@fynnfluegge Thanks for the work on this! PR looks good, but I want to test this myself first. Might take a few days before I get to it.
